lmfit/minimizer.py

```diff
@@ -499,6 +507,124 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
 
         return result
 
+    def emcee(self, params=None, steps=1000, nwalkers=100, burn=0, thin=1,
+              ntemps=1):
+        """
+        Bayesian sampling of the posterior distribution for the parameters
+        using the `emcee` Markov Chain Monte Carlo package.
+        The method assumes that the prior is Uniform.
+        You need to have `emcee` and `pandas` installed to use this method.
+
+        Parameters
+        ----------
+        steps : int, optional
+            How many samples you would like to draw from the posterior
+            distribution?
+        burn : int, optional
+            Discard this many samples from the start of the sampling regime.
+        thin : int, optional
+            Only accept 1 in every `thin` samples.
+        ntemps : int, optional
+            If `ntemps > 1` perform a Parallel Tempering.
+
+        Returns
+        -------
+        result, chain : MinimizerResult, pandas.DataFrame
+            MinimizerResult object contains updated params, fit statistics, etc.
+            The `chain` contains the samples. Has shape (steps, ndims). Access
+            chain values for a particular parameter using chain[parname].
+        """
+        if not HAS_EMCEE:
+            raise NotImplementedError('You must have emcee to use the emcee '
+                                      'method')
+
+        result = self.prepare_fit(params=params)
+        vars   = result.init_vals
+        params = result.params
+
+        # Removing internal parameter scaling. We could possibly keep it,
+        # but I don't know how this affects the emcee sampling.
+        bounds_varying = []
+        for i, par in enumerate(params):
+            param = params[par]
+            vars[i] = param.from_internal(param.value)
+            param.from_internal = lambda val: val
+            lb, ub = param.min, param.max
+            if lb is None or lb is np.nan:
+                lb = -np.inf
+            if ub is None or ub is np.nan:
+                ub = np.inf
+            bounds_varying.append((lb, ub))
+
+        bounds_varying = np.array(bounds_varying)
+
+        self.nvarys = len(result.var_names)
+
+        def lnlike(theta):
+            # log likelihood
+            return -0.5 * self.penalty(theta)
```

Comment:
The likelihood function was need to use is:

-\frac{1}{2}\sum_n \left[ \frac{(y_{obs} - y_{calc})^2}{s_n^2} + \ln (2\pi s_n^2) \right]
 (http://dan.iel.fm/emcee/current/user/line/)

lmfit instructs users to write an objective function that returns a (possibly weighted) residual.  The weighted residual vector is the first summand in the square brackets on the RHS.  The sum of the weighted residual vector is chi^2. chi^2 obviously varies as the model changes.
The second summand term,  \ln (2\pi s_n^2), is a constant and doesn't vary as parameters vary; unless one of the parameters is related to the noise term (as per the emcee example).
The `emcee` method only has access to `lmfit.penalty` or `lmfit.__residual`.  If I use `lmfit.penalty` (a scalar) I have to assume that it corresponds to the entirety of the likelihood function (assumes second summand is ignored), apart from multiplying it by -0.5.
If I use `lmfit.__residual` then the presumed behaviour should be:
* to square and sum if the output is a vector (this may under or overestimate depending on whether the second summand is included in the objective function output)
* to use as is if the output is a scalar
then to multiply by -0.5 as before.

One can get the exact likelihood function by saying that the user objective function has to return a scalar, which has included both summand terms correctly and has summed over all the points.

Unfortunately there is no way for the likelihood function to know what the objective function is returning (i.e. whether the user has weighted or not).